hazelcast/src/main/java/com/hazelcast/sql/impl/worker/QueryOperationWorker.java

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * Worker responsible for operation processing.
+ */
+public class QueryOperationWorker implements Runnable {
+
+    private static final Object POISON = new Object();
+
+    private final QueryOperationHandler operationHandler;
+    private final SerializationService ss;
+    private final Thread thread;
+    private final MPSCQueue<Object> queue;
+    private final ILogger logger;
+
+    private UUID localMemberId;
+
+    public QueryOperationWorker(
+        QueryOperationHandler operationHandler,
+        SerializationService ss,
+        String instanceName,
+        int index,
+        ILogger logger
+    ) {
+        this.operationHandler = operationHandler;
+        this.ss = ss;
+        this.logger = logger;
+
+        thread = new Thread(this,  QueryUtils.workerName(instanceName, WORKER_TYPE_OPERATION, index));
+        queue = new MPSCQueue<>(thread, null);
+
+        thread.start();
+    }
+
+    public void init(UUID localMemberId) {
+        this.localMemberId = localMemberId;
+    }
+
+    public void submit(QueryOperationExecutable task) {
+        queue.add(task);
+    }
+
+    public void stop() {
+        queue.clear();
+        queue.add(POISON);
+
+        thread.interrupt();
+    }
+
+    @Override
+    public void run() {
+        try {
+            run0();
+        } catch (Throwable t) {
+            inspectOutOfMemoryError(t);
+            logger.severe(t);
+        }
+    }
+
+    private void run0() {
+        try {
+            while (true) {
+                Object task = queue.take();
+
+                if (task == POISON) {
+                    break;
+                } else {
+                    assert task instanceof QueryOperationExecutable;
+
+                    execute((QueryOperationExecutable) task);
+                }
+            }
+        } catch (InterruptedException e) {
+            // No-op.
+        }
+    }
+
+    private void execute(QueryOperationExecutable task) {
+        QueryOperation operation = task.getLocalOperation();
+
+        if (operation == null) {
+            operation = deserialize(task.getRemoteOperation());
+        }
+
+        operationHandler.execute(operation);
+    }
+
+    /**
+     * Deserializes the packet into operation. If an exception happens, the query is cancelled.
+     *
+     * @param packet Packet packet.
+     * @return Query operation.
+     */
+    private QueryOperation deserialize(Packet packet) {
+        try {
+            return ss.toObject(packet);
+        } catch (Exception e) {
+            if (e.getCause() instanceof QueryOperationDeserializationException) {
+                QueryOperationDeserializationException error = (QueryOperationDeserializationException) e.getCause();
+
+                // We assume that only ID aware operations may hold user data. Other operations contain only HZ classes and
+                // we should never see deserialization errors for them.
+                sendDeserializationError(error);
```

Comment:
As part of the exception handling, we cancel the query here. In the else branch we don't do this. Is there any specific reason for this?

Reply:
The "else" branch is used for the situation where we cannot even extract the query ID (or the operation doesn't have query ID at all). Without it, we cannot cancel anything.

Reply:
Is there any other mechanism which will eventually cancel the query to avoid resources leak? If yes, I'd add this to the comment.

Reply:
No additional mechanisms in this specific PR.